### DISCUSSION TITLE
PorousFlowRelativePermeabilityConst for single phase flow

### DISCUSSION THREAD

**Discussion by ln53** [idaholab/moose#25716] [Q&A Modules: Porous Flow] [answered]
# Problem
I'm running single phase Porous Flow simulations and hoping to speed them up.

I thought I wouldn't need relative permeability in the single phase case, as phase disappearance shouldn't be a problem, and so tried a constant relative permeability of 1. This worked fine for a small 2D test and did actually reduce the run time, but as soon as I tried it on a more complex geometry, it would only get through one or two nonlinear iterations before failing to converge (`Linear solve did not converge due to DIVERGED_NANORINF iterations 0` -> `Nonlinear solve did not converge due to DIVERGED_LINEAR_SOLVE iterations 1`).

As a compromise I returned to the Corey version, and bumped `n` down to 2 (and even 1) as suggested in the page on [nonlinear convergence](https://mooseframework.inl.gov/modules/porous_flow/nonlinear_convergence_problems.html). This also failed to converge and only works with a minimum value of `n = 3`. I also (briefly) tried using FLAC, as it was faster than Corey on the simpler case, however that too failed on the more complex case.

What is likely to be causing this to happen, and is there a way I can get around it?

# Background
I'm actually simulating a fluid being injected into a completely dry porous medium, which isn't something directly supported by the Porous Flow module. I've written a custom "capillary pressure" user object, which calculates saturation on a curve between $S\approx0$ at $P=0$ and $S\approx1$ at $P=2P_f$ (where $P_f$ is some nominal pressure at the flow front).

This approach is working pretty well, and I've validated it against some analytical cases, however at the mesh resolutions I require (based on convergence studies with my material properties) the run times are far too long.

Hence I'm trying to simplify any aspects of Porous Flow not applicable to my case, relative permeability appearing to be a low hanging fruit. Incidentally if there are any other suggestions for simplifications/optimisations I could make, I'd love to hear them!

**Comment by GiudGiud** [comment]
Hello

Please try these general instructions for solving convergence issues.
Initializing better & picking direct solvers is likely going to help
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

Guillaume

**Reply by ln53** [reply]
Thanks for the link - I had followed some of the advice on that page already, this time I tried using `FDP` as suggested and it did indeed converge, apparently indicating the Jacobian needs to be "fixed".

To be honest I'm not sure why using a simpler relative permeability formulation would break the Jacobian, and even less sure how I'd go about fixing it, but I will do some more research and have a go. Absolute worst case I can use the slow (`n = 3`) version.

I don't think I can initialise any better, as my only variable (P) is set to a global IC of 0 which is the physical reality. I have also been running with the direct `lu` and `mumps` which hasn't helped.

**Reply by GiudGiud** [reply]
ok so if the Jacobian isnt great, did you try PJFNK instead of Newton?

**Reply by cpgr** [reply]
I'm pretty sure that the derivatives for the constant relative permeability are correct 😄. I suspect that when you use a non-constant relative permeability with the exponent ramped up, that the curve gets flatter, relatively perm gets smaller and the flow slows down significantly, making it easier to solve. 

You have your own capillary pressure - we do use the derivative of that in the Jacobian. Did you get them correct? And are you using the the PorousFlow1P material to compute saturations?

**Reply by ln53** [reply]
Thanks both for the input - PJFNK also has the same problem with constant/low order relative permeability unfortunately.

Okay that makes sense. I am using `PorousFlow1PhaseP` to compute saturation, and I'm pretty sure I've got the derivatives correct (checked on Wolfram and Desmos):
`effectiveSaturation` -> `1 / (1 + std::exp((pf - pc) / _epsilon))`
`dEffectiveSaturation` -> `(std::exp((pf - pc) / _epsilon)) / (_epsilon * std::pow(std::exp((pf - pc) / _epsilon) + 1, 2))`
`d2EffectiveSaturation` -> `(std::exp((2 * pf + pc) / _epsilon) - std::exp((pf + 2 * pc) / _epsilon)) / (std::pow(_epsilon, 2) * std::pow(std::exp(pf / _epsilon) + std::exp(pc / _epsilon), 3))`
Where `pc` is the pressure*, and `pf` is the nominal flow front pressure I mentioned above.

I should note that I've set `capillaryPressureCurve` and its derivatives to return 0 as I'll never use this object in a porepressure-saturation formulation.

\* I'm a little confused by this - afaik `pc` is supposed to denote capillary pressure, however in `PorousFlow1PhaseP` only the porepressure seems to be passed in to the saturation calcs (I had thought $P_c=-P$ in the 1 phase case). I don't think it helps that I'm working in absolute pressure and so negative pressure doesn't make much sense. Regardless, using the positive pressure value to calculate saturation has worked for me so far.

**Reply by ln53** [reply, marked as answer]
From this discussion it seems the runtime easy wins I had hoped for might not be so straightforward - thanks again for the help though.
